libcontainer/nsenter/nsexec.c

```diff
@@ -105,47 +110,64 @@ void nsexec()
 			exit(1);
 		}
 	}
-
-	/* Check that the specified process exists */
-	snprintf(buf, PATH_MAX - 1, "/proc/%d/ns", pid);
-	tfd = open(buf, O_DIRECTORY | O_RDONLY);
-	if (tfd == -1) {
-		pr_perror("Failed to open \"%s\"", buf);
-		exit(1);
+	// open all namespaces' descriptors and perform setns on them
+	nsLen = namespacesLength(nspaths);
+	if (nsLen == 0) {
+		return;
 	}
-
-	for (i = 0; i < num; i++) {
-		struct stat st;
-		int fd;
-
-		/* Symlinks on all namespaces exist for dead processes, but they can't be opened */
-		if (fstatat(tfd, namespaces[i], &st, AT_SYMLINK_NOFOLLOW) == -1) {
-			// Ignore nonexistent namespaces.
-			if (errno == ENOENT)
-				continue;
+	int fds[nsLen];
+	char *nsList[nsLen];
+	int i, j, savedErr = -1;
+	char *ns, *saveptr;
+	for (i = 0; i < nsLen; i++) {
+		ns = strtok_r(nspaths, ",", &saveptr);
+		if (ns == NULL) {
+			break;
 		}
-
-		fd = openat(tfd, namespaces[i], O_RDONLY);
-		if (fd == -1) {
-			pr_perror("Failed to open ns file %s for ns %s", buf,
-				  namespaces[i]);
+		fds[i] = open(ns, O_RDONLY);
+		if (fds[i] == -1) {
+			savedErr = errno;
```

Comment:
right, i think we also discussed before about this but i forgot to move the patch over. Since we exit anyway, it's ok to not having to close the fds IIRC, right ?